core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -531,8 +531,22 @@ assert equalType(
               final UnifyResult result = rule.apply(call);
               if (result != null) {
                 ++count;
-                attempted.add(new Replacement(result.call.query, result.result));
-                MutableRel parent = result.call.query.replaceInParent(result.result);
+                final MutableRel result1;
+                if (targetDescendant == target
+                    && !result.result.rowType.equals(result.call.query.rowType)) {
+                  // When target and query match with each other but have different
+                  // rowTypes, we add a MutableProject for compensation, thus to
+                  // assure the rowType of the replaced is exactly as before.
+                  final RexBuilder rexBuilder = target.cluster.getRexBuilder();
+                  List<RexNode> projects = new ArrayList<>(
+                      rexBuilder.identityProjects(result.call.query.rowType));
+                  result1 = MutableProject.of(
+                      result.call.query.rowType, result.result, projects);
+                } else {
+                  result1 = result.result;
+                }
+                attempted.add(new Replacement(result.call.query, result1));
+                MutableRel parent = result.call.query.replaceInParent(result1);
```

Comment:
`parent` variable is never used, we could get rid of it.

Reply:
Sure